README.md

A Flutter slider that makes a wave effect when dragged. Does a little bounce when dropped.

## Demo

<img height="240px" src="https://media.giphy.com/media/ifecPIFcXHu5AJmpfl/giphy.gif">


## Getting Started

To use this plugin, add `wave_slider` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).

The color can be set with the `color` property. The slider size is dependent on the size of its parent. The height of the wave slider can be set with the `sliderHeigth` property - which is constrained with a minimum of 50 and a maximum of 600.

An option `displayTrackball` property can be set to **true** to show a trackball along with the slider. The default is **false**

Values are retrieved by passing in an `onChanged` callback, which returns a value between 0 and 1 to indicate the current drag completion percentage.

The `onChangeStart` and `onChangeEnd` callbacks can be used to retrieve the start and end drag percentages respectively.

### Example

```dart
import 'package:flutter/material.dart';
import 'package:wave_slider/wave_slider.dart';

void main() => runApp(MaterialApp(
      home: App(),
    ));

class App extends StatefulWidget {
  @override
  _AppState createState() => _AppState();
}

class _AppState extends State<App> {
  double _dragPercentage = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          WaveSlider(
            displayTrackball: false,
            onChanged: (double dragUpdate) {
              setState(() {
                _dragPercentage = dragUpdate *
                    100; // dragUpdate is a fractional value between 0 and 1
              });
            },
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: Text(
              'Drag percentage',
              style: TextStyle(fontSize: 16, fontWeight: FontWeight.bold),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: Text(
              '$_dragPercentage',
              style: TextStyle(fontSize: 16),
            ),
          )
        ],
      ),
    );
  }
}
```

### Interested in how the package was made?
Check out this [playlist](https://www.youtube.com/playlist?list=PLjr4ufdmNA4J2-KwMutexAjjf_VmjL1eH) on the [Fun with Flutter](https://www.youtube.com/funwithflutter) YouTube channel!

CHANGELOG.md
## [0.2.0] - 23 March 2020

* Added an optional trackball to the slider that can be displayed by setting the `displayTrackball` property to **true**.

## [0.1.0] - 22 June 2019

* Ported from the original example. Working to return a fractional value between 0 and 1.
* Add safe area for paint.
* Add documentation and help.

test/wave_slider_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:wave_slider/wave_slider.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


example/example.dart
import 'package:flutter/material.dart';
import 'package:wave_slider/wave_slider.dart';

void main() => runApp(MaterialApp(
      home: App(),
    ));

class App extends StatefulWidget {
  @override
  _AppState createState() => _AppState();
}

class _AppState extends State<App> {
  double _dragPercentage = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          WaveSlider(
            onChanged: (double dragUpdate) {
              setState(() {
                _dragPercentage = dragUpdate *
                    100; // dragUpdate is a fractional value between 0 and 1
              });
            },
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: Text(
              'Drag percentage',
              style: TextStyle(fontSize: 16, fontWeight: FontWeight.bold),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: Text(
              '$_dragPercentage',
              style: TextStyle(fontSize: 16),
            ),
          )
        ],
      ),
    );
  }
}


lib/src/wave_slider.dart
import 'package:flutter/material.dart';
import 'package:wave_slider/src/wave_painter.dart';

class WaveSlider extends StatefulWidget {
  /// Creates a wave slider.
  ///
  /// When the state of the slider is changed the widget calls the [onChanged] callback.
  const WaveSlider({
    this.sliderHeight = 50.0,
    this.color = Colors.black,
    this.onChangeEnd,
    this.onChangeStart,
    @required this.onChanged,
    this.displayTrackball = false,
  })  : assert(onChanged != null),
        assert(color != null),
        assert(sliderHeight != null),
        assert(sliderHeight >= 50 && sliderHeight <= 600);

  /// The height of the slider can be set by specifying a [sliderHeight] - default is 50.0
  final double sliderHeight;

  /// The color of the slider can be set by specifying a [color] - default is black.
  final Color color;

  /// Called during a drag when the user is selecting a new value for the slider
  /// by dragging.
  ///
  /// Returns a percentage value between 0 and 100 for the current drag position.
  final ValueChanged<double> onChanged;

  /// Called when the user starts selecting a new value for the slider.
  final ValueChanged<double> onChangeStart;

  /// Called when the user is done selecting a new value for the slider.
  final ValueChanged<double> onChangeEnd;

  /// Display a trackball below the line on the current position as a visual indicator
  final bool displayTrackball;

  @override
  _WaveSliderState createState() => _WaveSliderState();
}

class _WaveSliderState extends State<WaveSlider>
    with SingleTickerProviderStateMixin {
  double _dragPosition = 0.0;
  double _dragPercentage = 0.0;
  double _sliderWidth = 0;

  WaveSliderController _slideController;

  @override
  void initState() {
    super.initState();
    _slideController = WaveSliderController(vsync: this)
      ..addListener(() => setState(() {}));
  }

  @override
  void dispose() {
    _slideController.dispose();
    super.dispose();
  }

  void _handleChanged(double val) {
    if (widget.onChanged != null) {
      widget.onChanged(val);
    }
  }

  void _handleChangeStart(double val) {
    if (widget.onChangeStart != null) {
      widget.onChangeStart(val);
    }
  }

  void _handleChangeEnd(double val) {
    if (widget.onChangeEnd != null) {
      widget.onChangeEnd(val);
    }
  }

  void _updateDragPosition(Offset val) {
    double newDragPosition = 0.0;
    if (val.dx <= 0.0) {
      newDragPosition = 0.0;
    } else if (val.dx >= _sliderWidth) {
      newDragPosition = _sliderWidth;
    } else {
      newDragPosition = val.dx;
    }

    setState(() {
      _dragPosition = newDragPosition;
      _dragPercentage = _dragPosition / _sliderWidth;
    });
  }

  void _onDragStart(BuildContext context, DragStartDetails start) {
    final RenderBox box = context.findRenderObject();
    final Offset localOffset = box.globalToLocal(start.globalPosition);
    _slideController.setStateToStart();
    _updateDragPosition(localOffset);
    _handleChangeStart(_dragPercentage);
  }

  void _onDragUpdate(BuildContext context, DragUpdateDetails update) {
    final RenderBox box = context.findRenderObject();
    final Offset localOffset = box.globalToLocal(update.globalPosition);
    _slideController.setStateToSliding();
    _updateDragPosition(localOffset);
    _handleChanged(_dragPercentage);
  }

  void _onDragEnd(BuildContext context, DragEndDetails end) {
    _slideController.setStateToStopping();
    setState(() {});
    _handleChangeEnd(_dragPercentage);
  }

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(
      builder: (BuildContext context, BoxConstraints constraints) {
        _sliderWidth = constraints.maxWidth;
        final double _safePadding = widget.sliderHeight / 4;
        return Padding(
          padding: EdgeInsets.only(
            left: 8.0,
            right: 8.0,
            top: _safePadding,
            bottom: widget.displayTrackball ? _safePadding * 2 : _safePadding,
          ),
          child: GestureDetector(
            child: Container(
              width: _sliderWidth,
              height: widget.sliderHeight,
              child: CustomPaint(
                painter: WavePainter(
                  color: widget.color,
                  sliderPosition: _dragPosition,
                  dragPercentage: _dragPercentage,
                  sliderState: _slideController.state,
                  animationProgress: _slideController.progress,
                  displayTrackball: widget.displayTrackball,
                ),
              ),
            ),
            onHorizontalDragStart: (DragStartDetails start) =>
                _onDragStart(context, start),
            onHorizontalDragUpdate: (DragUpdateDetails update) =>
                _onDragUpdate(context, update),
            onHorizontalDragEnd: (DragEndDetails end) =>
                _onDragEnd(context, end),
          ),
        );
      },
    );
  }
}

class WaveSliderController extends ChangeNotifier {
  WaveSliderController({@required TickerProvider vsync})
      : controller = AnimationController(vsync: vsync) {
    controller
      ..addListener(_onProgressUpdate)
      ..addStatusListener(_onStatusUpdate);
  }

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }

  final AnimationController controller;
  SliderState _state = SliderState.resting;

  void _onProgressUpdate() {
    notifyListeners();
  }

  void _onStatusUpdate(AnimationStatus status) {
    if (status == AnimationStatus.completed) {
      _onTransitionCompleted();
    }
  }

  void _onTransitionCompleted() {
    if (_state == SliderState.stopping) {
      setStateToResting();
    }
  }

  double get progress => controller.value;

  SliderState get state => _state;

  void _startAnimation() {
    controller.duration = Duration(milliseconds: 500);
    controller.forward(from: 0.0);
    notifyListeners();
  }

  void setStateToStart() {
    _startAnimation();
    _state = SliderState.starting;
  }

  void setStateToStopping() {
    _startAnimation();
    _state = SliderState.stopping;
  }

  void setStateToSliding() {
    _state = SliderState.sliding;
  }

  void setStateToResting() {
    _state = SliderState.resting;
  }
}

enum SliderState {
  starting,
  resting,
  sliding,
  stopping,
}


lib/src/wave_painter.dart
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:wave_slider/src/wave_slider.dart';

class WavePainter extends CustomPainter {
  WavePainter({
    @required this.sliderPosition,
    @required this.dragPercentage,
    @required this.animationProgress,
    @required this.sliderState,
    @required this.color,
    this.displayTrackball = false,
  })  : assert(sliderPosition != null &&
            dragPercentage != null &&
            animationProgress != null &&
            sliderState != null &&
            color != null),
        wavePainter = Paint()
          ..color = color
          ..style = PaintingStyle.stroke
          ..strokeWidth = 2.5,
        fillPainter = Paint()
          ..color = color
          ..style = PaintingStyle.fill;

  final double sliderPosition;
  final double dragPercentage;
  final double animationProgress;

  final SliderState sliderState;

  final bool displayTrackball;

  final Color color;

  final Paint wavePainter;
  final Paint fillPainter;

  /// Previous slider position initialised at the [anchorRadius], which is the start
  double _previousSliderPosition = anchorRadius;

  static const double anchorRadius = 5;

  double minWaveHeight;
  double maxWaveHeight;

  @override
  void paint(Canvas canvas, Size size) {
    final Size restrictedSize = Size(size.width - anchorRadius, size.height);
    _paintAnchors(canvas, restrictedSize);
    minWaveHeight = restrictedSize.height * 0.2;
    maxWaveHeight = restrictedSize.height * 0.8;
    switch (sliderState) {
      case SliderState.starting:
        _paintStartupWave(canvas, restrictedSize);
        break;
      case SliderState.resting:
        _paintRestingWave(canvas, restrictedSize);
        break;
      case SliderState.sliding:
        _paintSlidingWave(canvas, restrictedSize);
        break;
      case SliderState.stopping:
        _paintStoppingWave(canvas, restrictedSize);
        break;
      default:
        _paintSlidingWave(canvas, restrictedSize);
        break;
    }
  }

  void _paintAnchors(Canvas canvas, Size size) {
    canvas.drawCircle(
        Offset(anchorRadius, size.height), anchorRadius, fillPainter);
    canvas.drawCircle(
        Offset(size.width, size.height), anchorRadius, fillPainter);
  }

  void _paintRestingWave(Canvas canvas, Size size) {
    final Path path = Path();
    path.moveTo(anchorRadius, size.height);
    path.lineTo(size.width, size.height);
    canvas.drawPath(path, wavePainter);
    if (displayTrackball) {
      _paintTrackball(canvas, size);
    }
  }

  void _paintStartupWave(Canvas canvas, Size size) {
    final WaveCurveDefinitions line = _calculateWaveLineDefinitions(size);

    final double waveHeight = lerpDouble(size.height, line.controlHeight,
        Curves.elasticOut.transform(animationProgress));
    line.controlHeight = waveHeight;
    _paintWaveLine(canvas, size, line);
    if (displayTrackball) {
      _paintTrackball(canvas, size, waveCurve: line);
    }
  }

  void _paintSlidingWave(Canvas canvas, Size size) {
    final WaveCurveDefinitions line = _calculateWaveLineDefinitions(size);
    _paintWaveLine(canvas, size, line);
    if (displayTrackball) {
      _paintTrackball(canvas, size, waveCurve: line);
    }
  }

  void _paintStoppingWave(Canvas canvas, Size size) {
    final WaveCurveDefinitions line = _calculateWaveLineDefinitions(size);

    final double waveHeight = lerpDouble(line.controlHeight, size.height,
        Curves.elasticOut.transform(animationProgress));

    line.controlHeight = waveHeight;

    _paintWaveLine(canvas, size, line);
    if (displayTrackball) {
      _paintTrackball(canvas, size, waveCurve: line);
    }
  }

  void _paintWaveLine(
      Canvas canvas, Size size, WaveCurveDefinitions waveCurve) {
    final Path path = Path();
    path.moveTo(anchorRadius, size.height);
    path.lineTo(waveCurve.startOfBezier, size.height);
    path.cubicTo(
        waveCurve.leftControlPoint1,
        size.height,
        waveCurve.leftControlPoint2,
        waveCurve.controlHeight,
        waveCurve.centerPoint,
        waveCurve.controlHeight);
    path.cubicTo(
        waveCurve.rightControlPoint1,
        waveCurve.controlHeight,
        waveCurve.rightControlPoint2,
        size.height,
        waveCurve.endOfBezier,
        size.height);
    path.lineTo(size.width, size.height);

    canvas.drawPath(path, wavePainter);
  }

  void _paintTrackball(Canvas canvas, Size size,
      {WaveCurveDefinitions waveCurve}) {
    double indicatorSize = minWaveHeight;
    double centerPoint = sliderPosition, controlHeight = size.height;
    centerPoint = (centerPoint > size.width) ? size.width : centerPoint;
    if (waveCurve != null) {
      centerPoint = waveCurve.centerPoint;
      controlHeight = waveCurve.controlHeight;

      indicatorSize = (size.height - controlHeight) / 2.5;
      if (indicatorSize < minWaveHeight) {
        indicatorSize = minWaveHeight;
      }
    }
    canvas.drawCircle(Offset(centerPoint, controlHeight + indicatorSize * 1.5),
        indicatorSize, fillPainter);
  }

  WaveCurveDefinitions _calculateWaveLineDefinitions(Size size) {
    final double controlHeight =
        (size.height - minWaveHeight) - (maxWaveHeight * dragPercentage);

    final double bendWidth = 20 + 20 * dragPercentage;
    final double bezierWidth = 20 + 20 * dragPercentage;

    double centerPoint = sliderPosition;
    centerPoint = (centerPoint > size.width) ? size.width : centerPoint;

    double startOfBend = centerPoint - bendWidth / 2;
    double startOfBezier = startOfBend - bezierWidth;
    double endOfBend = sliderPosition + bendWidth / 2;
    double endOfBezier = endOfBend + bezierWidth;

    startOfBend = (startOfBend <= anchorRadius) ? anchorRadius : startOfBend;
    startOfBezier =
        (startOfBezier <= anchorRadius) ? anchorRadius : startOfBezier;
    endOfBend = (endOfBend > size.width) ? size.width : endOfBend;
    endOfBezier = (endOfBezier > size.width) ? size.width : endOfBezier;

    double leftBendControlPoint1 = startOfBend;
    double leftBendControlPoint2 = startOfBend;
    double rightBendControlPoint1 = endOfBend;
    double rightBendControlPoint2 = endOfBend;

    const double bendability = 25.0;
    const double maxSlideDifference = 30.0;
    double slideDifference = (sliderPosition - _previousSliderPosition).abs();

    slideDifference = (slideDifference > maxSlideDifference)
        ? maxSlideDifference
        : slideDifference;

    double bend =
        lerpDouble(0.0, bendability, slideDifference / maxSlideDifference);
    final bool moveLeft = sliderPosition < _previousSliderPosition;
    bend = moveLeft ? -bend : bend;

    leftBendControlPoint1 = leftBendControlPoint1 + bend;
    leftBendControlPoint2 = leftBendControlPoint2 - bend;
    rightBendControlPoint1 = rightBendControlPoint1 - bend;
    rightBendControlPoint2 = rightBendControlPoint2 + bend;

    centerPoint = centerPoint - bend;

    final WaveCurveDefinitions waveCurveDefinitions = WaveCurveDefinitions(
      controlHeight: controlHeight,
      startOfBezier: startOfBezier,
      endOfBezier: endOfBezier,
      leftControlPoint1: leftBendControlPoint1,
      leftControlPoint2: leftBendControlPoint2,
      rightControlPoint1: rightBendControlPoint1,
      rightControlPoint2: rightBendControlPoint2,
      centerPoint: centerPoint,
    );

    return waveCurveDefinitions;
  }

  @override
  bool shouldRepaint(WavePainter oldDelegate) {
    final double diff = _previousSliderPosition - oldDelegate.sliderPosition;
    if (diff.abs() > 20) {
      _previousSliderPosition = sliderPosition;
    } else {
      _previousSliderPosition = oldDelegate.sliderPosition;
    }
    return true;
  }
}

class WaveCurveDefinitions {
  WaveCurveDefinitions({
    @required this.startOfBezier,
    @required this.endOfBezier,
    @required this.leftControlPoint1,
    @required this.leftControlPoint2,
    @required this.rightControlPoint1,
    @required this.rightControlPoint2,
    @required this.controlHeight,
    @required this.centerPoint,
  });

  double startOfBezier;
  double endOfBezier;
  double leftControlPoint1;
  double leftControlPoint2;
  double rightControlPoint1;
  double rightControlPoint2;
  double controlHeight;
  double centerPoint;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:wave_slider/wave_slider.dart';

void main() => runApp(MaterialApp(
      home: App(),
    ));

class App extends StatefulWidget {
  @override
  _AppState createState() => _AppState();
}

class _AppState extends State<App> {
  double _dragPercentage = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          WaveSlider(
            displayTrackball: false,
            onChanged: (double dragUpdate) {
              setState(() {
                _dragPercentage = dragUpdate *
                    100; // dragUpdate is a fractional value between 0 and 1
              });
            },
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: Text(
              'Drag percentage',
              style: TextStyle(fontSize: 16, fontWeight: FontWeight.bold),
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: Text(
              '$_dragPercentage',
              style: const TextStyle(fontSize: 16),
            ),
          )
        ],
      ),
    );
  }
}
